AGP 9 forbids adding providers to srcDir

Testing AGP 9.0.0-alpha09 with your 5.6.8

```
Caused by: java.lang.reflect.InvocationTargetException
        at org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings.AndroidBinder.javaSrcDir(AndroidBinder.kt:158)
        at org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings.AndroidBinder.access$javaSrcDir(AndroidBinder.kt:21)
        at org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings.AndroidBinder$configure$2.execute(AndroidBinder.kt:47)
        at org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings.AndroidBinder$configure$2.execute(AndroidBinder.kt:38)
        at org.gradle.internal.code.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:124)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:109)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        ... 224 more
Caused by: java.lang.RuntimeException: Error : You cannot add Provider instances to the Android SourceSet API.
It is not possible for Android Studio to determine if the Provider points
to a directory that contains generated (read-only) or static (read-write) files.

Instead you should use the Sources interface in the Variant API, in particular
SourceDirectories.addGeneratedDirectory and SourceDirectories.addStaticDirectories.

You can re-enable the behavior by setting the `android.sourceset.disallowProvider=false` to gradle.properties.
However, be aware that any Gradle Task dependency will not be automatically carried.
        at com.android.build.gradle.internal.api.DefaultAndroidSourceDirectorySet.srcDir(DefaultAndroidSourceDirectorySet.kt:84)
        ... 233 more
```

--- a/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/BuildConfigPlugin.kt
+++ b/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/BuildConfigPlugin.kt
@@ -15,7 +15,6 @@ import org.gradle.api.plugins.PluginContainer
 import org.gradle.api.tasks.SourceSet
 import org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings.AndroidBinder
 import org.gradle.kotlin.dsl.domainObjectContainer
-import org.gradle.kotlin.dsl.newInstance
 import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
 import org.gradle.util.GradleVersion
@@ -151,10 +150,8 @@ class BuildConfigPlugin : Plugin<Project> {
             group = "BuildConfig"
             description = "Generates the build constants class for '${sourceSet.name}' source"
 
+            bindTo(sourceSet)
             outputDir.set(layout.buildDirectory.dir("generated/sources/buildConfig/${sourceSet.name}"))
-
-            specs.add(provider { isolate(sourceSet) })
-            specs.addAll(provider { sourceSet.extraSpecs.map { isolate(it) } })
         }
     }
 
@@ -170,25 +167,6 @@ class BuildConfigPlugin : Plugin<Project> {
         "org.jetbrains.kotlin.jvm",
         "org.jetbrains.kotlin.multiplatform",
     ).any(plugins::hasPlugin)
-
-    /**
-     * Helper method to create a copy of a BuildConfigClassSpec not linked to Gradle related extensions/DSL
-     * It makes it compatible with Configuration Cache
-     */
-    private fun Project.isolate(source: BuildConfigClassSpec) =
-        objects.newInstance<BuildConfigClassSpec>(source.name).apply spec@{
-            generator.value(source.generator).disallowChanges()
-            className.value(source.className).disallowChanges()
-            packageName.value(source.packageName).disallowChanges()
-            documentation.value(source.documentation).disallowChanges()
-            buildConfigFields.addAll(source.buildConfigFields.map { field ->
-                objects.newInstance<BuildConfigField>(field.name).apply field@{
-                    this@field.type.value(field.type).disallowChanges()
-                    this@field.value.value(field.value).disallowChanges()
-                    this@field.position.value(field.position).disallowChanges()
-                }
-            })
-        }
 
     private val Project.defaultPackage
         get() = provider {

--- a/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/BuildConfigTask.kt
+++ b/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/BuildConfigTask.kt
@@ -3,8 +3,10 @@
 package com.github.gmazzo.buildconfig
 
 import com.github.gmazzo.buildconfig.generators.BuildConfigGeneratorSpec
+import com.github.gmazzo.buildconfig.internal.BuildConfigSourceSetInternal
 import java.io.File
 import org.gradle.api.DefaultTask
+import org.gradle.api.Project
 import org.gradle.api.Task
 import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.provider.SetProperty
@@ -13,6 +15,7 @@ import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.Nested
 import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.newInstance
 import org.gradle.util.GradleVersion
 
 @CacheableTask
@@ -65,6 +68,30 @@ abstract class BuildConfigTask : DefaultTask() {
             )
         }
     }
+
+    internal fun bindTo(sourceSet: BuildConfigSourceSetInternal) = with(project) {
+        specs.add(provider { isolate(sourceSet) })
+        specs.addAll(provider { sourceSet.extraSpecs.map { isolate(it) } })
+    }
+
+    /**
+     * Helper method to create a copy of a BuildConfigClassSpec not linked to Gradle related extensions/DSL
+     * It makes it compatible with Configuration Cache
+     */
+    private fun Project.isolate(source: BuildConfigClassSpec) =
+        objects.newInstance<BuildConfigClassSpec>(source.name).apply spec@{
+            generator.value(source.generator).disallowChanges()
+            className.value(source.className).disallowChanges()
+            packageName.value(source.packageName).disallowChanges()
+            documentation.value(source.documentation).disallowChanges()
+            buildConfigFields.addAll(source.buildConfigFields.map { field ->
+                objects.newInstance<BuildConfigField>(field.name).apply field@{
+                    this@field.type.value(field.type).disallowChanges()
+                    this@field.value.value(field.value).disallowChanges()
+                    this@field.position.value(field.position).disallowChanges()
+                }
+            })
+        }
 
     private object HasFields : Spec<Task> {
         override fun isSatisfiedBy(task: Task): Boolean =

--- a/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/internal/bindings/AndroidBinder.kt
+++ b/plugin/src/main/kotlin/com/github/gmazzo/buildconfig/internal/bindings/AndroidBinder.kt
@@ -2,16 +2,15 @@ package org.gradle.kotlin.dsl.com.github.gmazzo.buildconfig.internal.bindings
 
 import com.github.gmazzo.buildconfig.BuildConfigExtension
 import com.github.gmazzo.buildconfig.BuildConfigTask
+import com.github.gmazzo.buildconfig.internal.BuildConfigSourceSetInternal
 import com.github.gmazzo.buildconfig.internal.bindings.JavaBinder.registerExtension
 import org.gradle.api.Action
 import org.gradle.api.Named
 import org.gradle.api.NamedDomainObjectContainer
 import org.gradle.api.Project
 import org.gradle.api.Task
-import org.gradle.api.file.Directory
 import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.plugins.ExtensionAware
-import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.SourceSet.MAIN_SOURCE_SET_NAME
 import org.gradle.api.tasks.SourceSet.TEST_SOURCE_SET_NAME
 import org.gradle.api.tasks.TaskProvider
@@ -42,41 +41,41 @@ internal object AndroidBinder {
             val spec = extension.sourceSets.maybeCreate(kmpAwareName)
 
             (it as ExtensionAware).registerExtension(spec)
-            it.javaSrcDir(spec.generateTask.flatMap { it.outputDir })
         }
 
         androidComponentsOnVariants { variant ->
-            val unitTest = variant.unitTest
-            val androidTest = variant.androidTest
+            variant.bindToSourceSet(extension,
+                mainSourceSetName,
+                variant.buildType,
+                variant.flavorName,
+                *variant.productFlavors.map { (_, flavor) -> flavor }.toTypedArray())
 
-            for (component in listOfNotNull(variant, unitTest, androidTest)) {
-                val specNames = when (component) {
-                    variant -> sequenceOf(
-                        mainSourceSetName,
-                        component.name,
-                        component.buildType,
-                        component.flavorName
-                    ) + component.productFlavors.asSequence().map { (_, flavor) -> flavor }
+            variant.unitTest?.bindToSourceSet(extension, testSourceSetName)
+            variant.androidTest?.bindToSourceSet(extension, "androidTest")
+        }
+    }
 
-                    unitTest -> sequenceOf(
-                        testSourceSetName,
-                        component.name
-                    )
+    private fun Any/*Component*/.bindToSourceSet(
+        extension: BuildConfigExtension,
+        vararg extras: String?,
+    ) {
+        val ss = extension.sourceSets.maybeCreate(this@bindToSourceSet.name!!)
+        val extraSSs = extras.asSequence()
+            .filterNotNull()
+            .filter { it.isNotBlank() && it != ss.name }
+            .distinct()
+            .map(extension.sourceSets::maybeCreate)
 
-                    androidTest -> sequenceOf("androidTest", component.name)
-                    else -> error("Unsupported component $component")
-                }
-                val specs = specNames
-                    .filterNotNull()
-                    .filter(String::isNotBlank)
-                    .toSet()
-                    .map(extension.sourceSets::maybeCreate)
-
-                for (spec in specs) {
-                    component.sourcesJavaAddGeneratedSourceDirectory(spec.generateTask, BuildConfigTask::outputDir)
-                }
+        for (extraSS in extraSSs) {
+            extraSS.generateTask.configure {
+                it.onlyIf("It was superseded by ${ss.name} source set") { false }
+            }
+            ss.generateTask.configure {
+                it.bindTo(extraSS as BuildConfigSourceSetInternal)
             }
         }
+
+        sourcesJavaAddGeneratedSourceDirectory(ss.generateTask, BuildConfigTask::outputDir)
     }
 
     // project.androidComponents.onVariants
@@ -148,13 +147,5 @@ internal object AndroidBinder {
             javaClass.getMethod("getSourceSets")
                 .invoke(this) as NamedDomainObjectContainer<Named>
         }
-
-    // AndroidSourceSet.java.srcDir(dir)
-    private fun Named.javaSrcDir(dir: Provider<Directory>) {
-        with(javaClass.getMethod("getJava").invoke(this)) {
-            javaClass.getMethod("srcDir", Any::class.java)
-                .invoke(this, dir)
-        }
-    }
 
 }

